File: app/src/main/java/app/fit/fitndflow/ui/features/exercises/ExerciseListFragment.java
package app.fit.fitndflow.ui.features.exercises;


import static app.fit.fitndflow.ui.features.categories.CreationOrModifyInputDialog.TYPE_EXERCISE;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.fit.fitndflow.R;
import com.fit.fitndflow.databinding.FragmentExercisesListBinding;

import java.util.ArrayList;
import java.util.List;

import app.fit.fitndflow.domain.model.CategoryModel;
import app.fit.fitndflow.domain.model.ExerciseModel;
import app.fit.fitndflow.ui.features.categories.CategoriesAndExercisesViewModel;
import app.fit.fitndflow.ui.features.categories.ConfirmationDialogFragment;
import app.fit.fitndflow.ui.features.categories.CreationOrModifyInputDialog;
import app.fit.fitndflow.ui.features.categories.DialogCallbackDelete;
import app.fit.fitndflow.ui.features.common.CommonFragment;
import app.fit.fitndflow.ui.features.training.AddSerieTrainingFragment;
import app.fit.fitndflow.ui.features.training.SerieAdapterCallback;

public class ExerciseListFragment extends CommonFragment implements SerieAdapterCallback, DialogCallbackDelete {

    private FragmentExercisesListBinding binding;

    private CategoriesAndExercisesViewModel categoriesAndExercisesViewModel;

    private ExercisesAdapter exercisesAdapter;
    private boolean isEditMode;

    @Override
    protected Class getViewModelClass() {
        return CategoriesAndExercisesViewModel.class;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentExercisesListBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        super.onCreateView(inflater, container, savedInstanceState);
        categoriesAndExercisesViewModel = ViewModelProviders.of(getActivity()).get(CategoriesAndExercisesViewModel.class);
        printCategoryDetailAndinstantiateAdaper(categoriesAndExercisesViewModel.getActualCategory().getValue());
        setViewModelObservers();
        setOnClickListeners();
        addTextWatcher();
        return view;
    }

    private void setViewModelObservers() {
        categoriesAndExercisesViewModel.getMutableSlideError().setValue(false);
        categoriesAndExercisesViewModel.getIsLoading().setValue(false);

        final Observer<CategoryModel> observer = new Observer<CategoryModel>() {
            @Override
            public void onChanged(CategoryModel category) {
                //setear listado al adapter
                exercisesAdapter.setExerciseModelList(category.getExerciseList());
            }
        };
        categoriesAndExercisesViewModel.getActualCategory().observe(getActivity(), observer);

        final Observer<Boolean> observerError = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isError) {
                if (isError) {
                    showSlideError();
                    categoriesAndExercisesViewModel.getMutableSlideError().setValue(false);
                }
            }
        };
        categoriesAndExercisesViewModel.getMutableSlideError().observe(getActivity(), observerError);
        final Observer<Boolean> errorObserver = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isError) {
                if (isError) {
                    printError();
                    categoriesAndExercisesViewModel.getMutableFullScreenError().setValue(false);
                }
            }
        };
        categoriesAndExercisesViewModel.getMutableFullScreenError().observe(getActivity(), errorObserver);
        final Observer<Boolean> observerIsSaveSuccess = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isSaveSuccess) {
                if (isSaveSuccess) {
                    showSlideSaved();
                    categoriesAndExercisesViewModel.getIsSaveSuccess().setValue(false);
                }
            }
        };
        categoriesAndExercisesViewModel.getIsSaveSuccess().observe(getActivity(), observerIsSaveSuccess);
        final Observer<Boolean> observerLoading = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isLoading) {
                try {
                    if (isLoading) {
                        showLoading();

                    } else {
                        hideLoading();
                    }
                } catch (Exception exception) {
                    Log.e("Error", "show loading");
                }
            }
        };
        categoriesAndExercisesViewModel.getIsLoading().observe(getActivity(), observerLoading);
    }

    private void printCategoryDetailAndinstantiateAdaper(CategoryModel categoryRecived) {
        exercisesAdapter = new ExercisesAdapter(categoryRecived.getExerciseList(), this);
        binding.recyclerCategories.setHasFixedSize(true);
        binding.recyclerCategories.setLayoutManager(new LinearLayoutManager(this.getContext()));
        binding.recyclerCategories.setAdapter(exercisesAdapter);

        binding.categoryTitle.setText(categoryRecived.getName());
    }

    private void addTextWatcher() {
        binding.txtSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                List<ExerciseModel> filteredList = new ArrayList<>();
                CategoryModel category = categoriesAndExercisesViewModel.getActualCategory().getValue();
                List<ExerciseModel> exerciseList = category.getExerciseList();
                for (int j = 0; j < exerciseList.size(); j++) {
                    if (exerciseList.get(j).getName().toUpperCase().contains(charSequence.toString().toUpperCase())) {
                        filteredList.add(exerciseList.get(j));
                    }
                }
                exercisesAdapter.setExerciseModelList(filteredList);
            }
            @Override
            public void afterTextChanged(Editable editable) {

            }
        });
    }
    private void setOnClickListeners() {
        binding.floatingBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                isEditMode = !isEditMode;
                if (isEditMode) {
                    binding.floatingBtn.setImageResource(R.drawable.svg_check);
                } else {
                    binding.floatingBtn.setImageResource(R.drawable.svg_pencil);
                }
                exercisesAdapter.setEditMode(isEditMode);
            }
        });
    }

    @Override
    public void showSeries(ExerciseModel exercise) {
        if (exercise.getId() != 0 && exercise.getName() != null) {
            addFragment(AddSerieTrainingFragment.newInstance(exercise));
        } else {
            showBlockError();
        }
    }
    private void printError() {
        try {
            showBlockError();
        } catch (Exception exception) {
            Log.e("Error", "Error to print errorContainer");
        }
    }
    @Override
    public void showCreationDialog() {
        CreationOrModifyInputDialog.newInstance(TYPE_EXERCISE).show(getChildFragmentManager(), "creationDialog");
    }

    @Override
    public void showDeleteDialog(int id) {
        ConfirmationDialogFragment.newInstance(ExerciseListFragment.this, ConfirmationDialogFragment.DELETE_EXERCISE, id).show(getChildFragmentManager(), "ConfirmationDialog");
    }

    @Override
    public void showModifyDialog(int exerciseId, String exerciseName) {
        CreationOrModifyInputDialog.newInstance(TYPE_EXERCISE, exerciseId, exerciseName).show(getChildFragmentManager(), "creationDialog");
    }

    @Override
    public void onClickAcceptDelete(int exerciseId) {
        categoriesAndExercisesViewModel.deleteExercise(exerciseId,requireContext());
    }
}

File: app/src/main/java/app/fit/fitndflow/ui/features/exercises/ExerciseListFragment.kt
package app.fit.fitndflow.ui.features.exercises

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import app.fit.fitndflow.domain.model.CategoryModel
import app.fit.fitndflow.domain.model.ExerciseModel
import app.fit.fitndflow.ui.features.categories.ConfirmationDialogFragment
import app.fit.fitndflow.ui.features.categories.CreationOrModifyInputDialog
import app.fit.fitndflow.ui.features.categories.CreationOrModifyInputDialog.Companion.TYPE_EXERCISE
import app.fit.fitndflow.ui.features.categories.DialogCallbackDelete
import app.fit.fitndflow.ui.features.common.CommonFragment
import app.fit.fitndflow.ui.features.training.AddSerieTrainingFragment
import app.fit.fitndflow.ui.features.training.SerieAdapterCallback
import com.fit.fitndflow.R
import com.fit.fitndflow.databinding.FragmentExercisesListBinding
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

class ExerciseListFragment : CommonFragment(), SerieAdapterCallback, DialogCallbackDelete {

    companion object {
        val KEY_CATEGORY = "actualCategory"
        fun newInstance(category: CategoryModel): ExerciseListFragment {
            val exerciseListFragment = ExerciseListFragment()
            val bundle = Bundle()
            bundle.putSerializable(KEY_CATEGORY, category)
            exerciseListFragment.arguments = bundle
            return exerciseListFragment
        }
    }

    private lateinit var binding: FragmentExercisesListBinding
    private lateinit var exercisesAdapter: ExercisesAdapter
    private val exercisesViewModel: ExercisesViewModel by activityViewModels()
    private val category: CategoryModel by lazy { requireArguments().getSerializable(KEY_CATEGORY) as CategoryModel }
    private var isEditMode: Boolean = false

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentExercisesListBinding.inflate(layoutInflater)
        val myView = binding.root
        super.onCreateView(inflater, container, savedInstanceState)
        setOnClickListeners()
        addTextWatcher()
        return myView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        attachObservers()
        binding.categoryTitle.setText(category.name)
        category.exerciseList?.let { instantiateExercisesAdapter(it) }
    }

    private fun attachObservers() {
        exercisesViewModel.state.onEach(::handleState).launchIn(viewLifecycleOwner.lifecycleScope)
    }

    private fun handleState(state: ExercisesViewModel.State) {
        when (state) {
            is ExercisesViewModel.State.ExerciseListRecived -> {
                printExercises(state.exerciseList.toMutableList())
                showSlideSaved()
                hideLoading()
            }

            ExercisesViewModel.State.SlideError -> {
                hideLoading()
                showSlideError()
            }

            ExercisesViewModel.State.Loading -> {
                showLoading()
            }
        }

//        exercisesViewModel.mutableSlideError.value = false
//        exercisesViewModel.isLoading.value = false
//
//        val categoryObserver = Observer<CategoryModel> { category ->
//            exercisesAdapter.setExerciseModelList(category.exerciseList)
//        }
//        exercisesViewModel.actualCategory.observe(viewLifecycleOwner, categoryObserver)
//
//        val slideErrorObserver = Observer<Boolean> { isError ->
//            if (isError) {
//                showSlideError()
//                exercisesViewModel.mutableSlideError.value = false
//            }
//        }
//        exercisesViewModel.mutableSlideError.observe(
//            viewLifecycleOwner,
//            slideErrorObserver
//        )
//
//        val fullScreenErrorObserver = Observer<Boolean> { isError ->
//            if (isError) {
//                showBlockError()
//                exercisesViewModel.mutableFullScreenError.value = false
//            }
//        }
//        exercisesViewModel.mutableFullScreenError.observe(
//            viewLifecycleOwner,
//            fullScreenErrorObserver
//        )
//
//        val savedSuccessOberver = Observer<Boolean> { isSaved ->
//            if (isSaved) {
//                showSlideSaved()
//                exercisesViewModel.isSaveSuccess.value = false
//            }
//        }
//        exercisesViewModel.isSaveSuccess.observe(
//            viewLifecycleOwner,
//            savedSuccessOberver
//        )
//
//        val observerLoading = Observer<Boolean> { isLoading ->
//            if (isLoading) {
//                showLoading()
//            } else {
//                hideLoading()
//            }
//        }
//        exercisesViewModel.isLoading.observe(viewLifecycleOwner, observerLoading)
    }

    private fun printExercises(exerciseListRecived: MutableList<ExerciseModel>){
       category.exerciseList = exerciseListRecived
       exercisesAdapter.setExerciseModelList(category.exerciseList)
    }

    private fun instantiateExercisesAdapter(exerciseListRecived: List<ExerciseModel>) {
        exercisesAdapter = ExercisesAdapter(exerciseListRecived, this)
        binding.recyclerCategories.setHasFixedSize(true)
        binding.recyclerCategories.layoutManager = LinearLayoutManager(this.context)
        binding.recyclerCategories.adapter = exercisesAdapter
    }

    private fun addTextWatcher() {
        binding.txtSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
                //Nothing to do
            }

            override fun onTextChanged(charSequence: CharSequence?, p1: Int, p2: Int, p3: Int) {
                charSequence?.let { searchText ->
                    val filteredList = mutableListOf<ExerciseModel>()
                    val exerciseList: List<ExerciseModel> = category.exerciseList!!
                    exerciseList.forEach { exercise ->
                        if (exercise.name.contains(searchText, ignoreCase = true)) {
                            filteredList.add(exercise)
                        }
                    }
                    exercisesAdapter.setExerciseModelList(filteredList)
                }
            }

            override fun afterTextChanged(p0: Editable?) {
                //Nothing to do
            }
        })
    }

    private fun setOnClickListeners() {
        binding.apply {
            floatingBtn.setOnClickListener {
                isEditMode = !isEditMode
                if (isEditMode) {
                    floatingBtn.setImageResource(R.drawable.svg_check)
                } else {
                    floatingBtn.setImageResource(R.drawable.svg_pencil)
                }
                exercisesAdapter.setEditMode(isEditMode)
            }
        }
    }

    override fun onClickAcceptDelete(exerciseId: Int) {         //todo!!!!!!!!!!!!!!!!!!!!
       // exercisesViewModel.deleteExercise(exerciseId, requireContext())
    }

    override fun showSeries(exercise: ExerciseModel) {
        if (exercise.id != 0) {
            addFragment(AddSerieTrainingFragment.newInstance(exercise))
        } else {
            showBlockError()
        }
    }

    override fun showCreationDialog() {
        CreationOrModifyInputDialog.newInstance(TYPE_EXERCISE, category.id)
            .show(childFragmentManager, CreationOrModifyInputDialog.TAG)
    }

    override fun showDeleteDialog(id: Int) {
        ConfirmationDialogFragment.newInstance(this, ConfirmationDialogFragment.DELETE_EXERCISE, id)
            .show(childFragmentManager, ConfirmationDialogFragment.TAG)
    }

    override fun showModifyDialog(exerciseId: Int, exerciseName: String) {
        CreationOrModifyInputDialog.newInstance(TYPE_EXERCISE, exerciseId, exerciseName, category.id)
            .show(childFragmentManager, CreationOrModifyInputDialog.TAG)
    }
}